### DISCUSSION TITLE
I'm really confused, HTTP posting to FastAPI using C#

### DISCUSSION THREAD

**Discussion by Todd82TA** [fastapi/fastapi#6015] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
I don't have any example code, that's what I'm hoping someone can share with me.
```

### Description

It's been nearly a decade since I've developed last. I am using Visual Studio 2022 and a lot has changed since I used VS2008. I am trying to develop a Windows service "agent" that will submit (on a timer), information to FastAPI, which is currently running on a Linux server.

For the life of me, I cannot figure out how to do this. All the documentation online that I see appears to be so centrally focused on either submitting information from the server to the agent, or doing things on the server. I've seen nothing related to how to submit information from a CLIENT (in this case, written in C#) to the server API. 

I know how to send an HTTP request (or post) in C#, but no idea what I should be doing with respect to Fast API. I would really appreciate any help or advice that anyone can give me. I feel like I'm missing the big picture here because everything I see is written in Python, and I'm doing this in C#. Why do I not see any example code anywhere for submitting TO the API from a client?

Thank you so much for anyone who is willing to help me...

### Operating System

Windows

### FastAPI Version

Latest

### Python Version

None

**Comment by mr-st0rm** [comment, marked as answer]
you can make a separate router for your client and do ip filtering if the client is on a static server(something like that)

and on the FastApi side to do the necessary actions

**Reply by Todd82TA** [reply]
Thank you so much for the response! I want to clarify to make sure I'm not confusing. Basically, yes, we have FastAPI set-up on a server, which will have a static IP. We'll have several clients which will submit data from the client machine back to the server. We don't (at this point) have any data that needs to go back from the server to the client.

The documentation that I look up for FastAPI seems to be really confusing, or maybe I'm just not understanding this properly. My understanding of how I would use an API (incidentally, from over a decade ago), would be that if I was connecting to it remotely via the web, I'd usually pass a string through the URL. I realize now we're encrypting traffic so that stuff doesn't get passed in open channels... but I'm confused as to how this information gets passed.

Conceivably, I should be able to use an HTTP request to communicate with the FastAPI API, right? I shouldn't need to include any libraries from FastAPI on the client side? I was asked to use C# simply because the clients are all going to be Windows-based... but I'm not seeing any examples, anywhere, of FastAPI being used with clients, with the router being developed in C#. 

I posted on Stack Exchange, and they closed / deleted my question because they thought I was shilling for FastAPI I guess... it was confusing at best. I simply was asking if anyone had any example client-based code in C# to interact with FastAPI on the server. 

Hope that helps, I apologize... I'm completely lost.

**Reply by jgould22** [reply]
I do not have any experience with windows, nor its service agent architectures.

>  My understanding of how I would use an API (incidentally, from over a decade ago), would be that if I was connecting to it remotely via the web, I'd usually pass a string through the URL.

[REST](https://en.wikipedia.org/wiki/Representational_state_transfer) semantics have been the same for some time and using query parameters is one option when asking for data from the server.  Here is a quick primer on REST semantics https://www.codecademy.com/article/what-is-rest

> I realize now we're encrypting traffic so that stuff doesn't get passed in open channels... but I'm confused as to how this information gets passed.

TLS/SSL (encrypting traffic) is **not a default** and requires explicit configuration. FastAPI does not really support that easily out of the box so keep that in mind.

> I've seen nothing related to how to submit information from a CLIENT (in this case, written in C#) to the server API.

To communicate with the FastAPI code you would use HTTP requests (FastAPI also provides some other ways like websockets but stick to RESTful HTTP requests for now) as you have mentioned. Usually if you are sending data from a client it is done with HTTP requests that contain JSON documents in the body.

HTTP Requests are not hard to make in C#(dotnet), Microsoft has created a tutorial exclusively for that here https://learn.microsoft.com/en-us/dotnet/csharp/tutorials/console-webapiclient which only does a get but there is more documentation about the other verbs here https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient
 
Best of luck

**Reply by Todd82TA** [reply]
Thank you, I really appreciate it. I had that come to Jesus moment a few minutes ago too. I didn't realize I could just inherently use the httpclient post/get to communicate with it. Everything I had seen showed adding a library in Python, and I just got wrapped around the wheel with that. Thank you!!!
